llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp

```diff
@@ -306,7 +306,7 @@ class CallsiteContextGraph {
     // True if this node was effectively removed from the graph, in which case
     // its context id set, caller edges, and callee edges should all be empty.
     bool isRemoved() const {
-      assert(ContextIds.empty() ==
+      assert(!ContextIds.empty() ||
```

Comment:
An assert with an or condition is a little hard to reason about for me. For this to fail both LHS and RHS need to be false - context IDs is empty *and* at least one of CalleeEdges, CallerEdges is not empty. Can we rewrite this as two separate assertions to simplify? Also it seems that it is now inconsistent with the comment above?

Reply:
The original assert was wrong, which showed up in testing after the reorganization here. The case that was wrong was when we were left with a single (non-removed) node in the graph, which would have non-empty context ids, but no callee or caller edges. I played around with restructuring the assertion but didn't come up with anything that looked clearer to me. E.g. here's another structure but it feels more convoluted to me personally:

```
assert(!(ContextIds.empty() &&
               (!CalleeEdges.empty() || !CallerEdges.empty()));
```

It's not inconsistent with the comment (it is still true that all 3 should be empty if it was removed) but I've added new comments to explain the case described above. Let me know if you have a preference of what to do for the assert structure.

Reply:
It is equivalent, it just means an extra check that is executed unnecessarily. Although hopefully the compiler would common it with the check on the return?

Reply:
I dont think there will be an extra check. With clang for non-assert builds (where NDEBUG is defined) the IR has an empty basic block [1] which should get optimized away in the generated code.

[1] https://godbolt.org/z/r5fceqeTK

Reply:
Good point, will go ahead and restructure like that.